File: java/src/main/java/br/com/emendes/problems/LongestValidParentheses.java
package br.com.emendes.problems;

/**
 * Dada uma string contendo apenas os caracteres '(' e ')', retorne o comprimento da maior substring que é um
 * parêntese válido (bem formado).<br><br>
 * <p>
 * Restrições:<br>
 * 0 <= s.length <= 3 * 10⁴<br>
 * s[i] é '(', or ')'.
 */
public class LongestValidParentheses {

  // Complexidade de tempo O(n)
  // Complexidade de espaço O(n)
  public int longestValidParentheses(String s) {
    int sLength = s.length();

    if (sLength < 2) return 0;

    int[] cache = new int[sLength];
    int answer = 0;

    if (s.charAt(0) == '(' && s.charAt(1) == ')') {
      cache[1] = 2;
      answer = 2;
    }

    for (int index = 2; index < sLength; index++) {
      char before = s.charAt(index - 1);
      char actual = s.charAt(index);
      if (actual == ')') {
        if (before == '(') {
          cache[index] = cache[index - 2] + 2;
          if (cache[index] > answer)
            answer = cache[index];
        } else {
          int i = index - cache[index - 1] - 1; //índice que deve conter um '(' para o index atual ser um parênteses válido.
          if (i > -1 && s.charAt(i) == '(') {
            i -= 1;
            cache[index] = cache[index - 1]  + 2;
            if (i > -1) {
              cache[index] += cache[i];
            }
            if (cache[index] > answer) {
              answer = cache[index];
            }
          }
        }
      }
    }

    return answer;
  }

}

File: javascript/problems/LongestValidParentheses.js
/**
 * Dado uma string contendo apenas caracteres '(' e ')', retorne o comprimento da maior substring de parênteses
 * bem formado.
 *
 * Restrições:
 *
 * 0 <= s.length <= 3 * 10⁴
 *
 * s[i] é '(', or ')'.
 *
 * @param {string} s
 * @return {number}
 */
var longestValidParentheses = function (s) {
  const open = [];
  const previousLength = [];
  let maxLength = 0;
  for (let i = 0; i < s.length; i++) {
    if (s[i] === "(") open.push(i);
    else if (open.length) {
      let start = open.pop();
      let currLength = i - start + 1 + (previousLength[start] || 0);
      maxLength = Math.max(maxLength, currLength);
      previousLength[i + 1] = currLength;
    }
  }

  return maxLength;
};
